Transformations/Transform/Transform.cpp

```diff
@@ -0,0 +1,340 @@
+
+        rotm.at<double>(0, 0) = 1 - 2 * pow(qy, 2) - 2 * pow(qz, 2);
+        rotm.at<double>(0, 1) = 2 * qx * qy - 2 * qz * qw;
+        rotm.at<double>(0, 2) = 2 * qx * qz + 2 * qy * qw;
+        rotm.at<double>(1, 0) = 2 * qx * qy + 2 * qz * qw;
+        rotm.at<double>(1, 1) = 1 - 2 * pow(qx, 2) - 2 * pow(qz, 2);
+        rotm.at<double>(1, 2) = 2 * qy * qz - 2 * qx * qw;
+        rotm.at<double>(2, 0) = 2 * qx * qz - 2 * qy * qw;
+        rotm.at<double>(2, 1) = 2 * qy * qz + 2 * qx * qw;
+        rotm.at<double>(2, 2) = 1 - 2 * pow(qx, 2) - 2 * pow(qy, 2);
+
+        return rotm;
+    }
+
+    cv::Mat rotm2quat(cv::Mat &rotm)
+    {
+        cv::Mat quat(4, 1, CV_64FC1, cv::Scalar(0));
+
+        quat.at<double>(0) = sqrt(1 + rotm.at<double>(0, 0) + rotm.at<double>(1, 1) + rotm.at<double>(2, 2)) / 2;
+        quat.at<double>(1) = (rotm.at<double>(2, 1) - rotm.at<double>(1, 2)) / (4 * quat.at<double>(0));
+        quat.at<double>(2) = (rotm.at<double>(0, 2) - rotm.at<double>(2, 0)) / (4 * quat.at<double>(0));
+        quat.at<double>(3) = (rotm.at<double>(1, 0) - rotm.at<double>(0, 1)) / (4 * quat.at<double>(0));
+
+        return quat;
+    }
+
+    std::tuple<double, cv::Mat> rotmat2angleaxis(cv::Mat &rotm)
+    {
+        cv::Mat rotvec = cv::Mat(3, 1, CV_64FC1, cv::Scalar(0));
+        cv::Rodrigues(rotm, rotvec);
+        double angle = cv::norm(rotvec);
+        cv::Mat axis = rotvec / cv::norm(rotvec);
+
+        return std::make_tuple(angle, axis);
+    }
+
+    cv::Mat rotvec2quat(cv::Mat rotvec, double angle = nan(""))
+    {
+        if(isnan(angle))
+        {
+            angle = cv::norm(rotvec);
+        }
+
+        cv::Mat quat(4, 1, CV_64FC1, cv::Scalar(0));
+        quat.at<double>(0) = cos(0.5 * angle);
+        if(quat.at<double>(0) == 0)
+        {
+            throw "cos(0.5 * angle) = 0";
+        }
+        auto a = sqrt((1 - pow(quat.at<double>(0), 2)) / pow(angle, 2));
+        for(int i = 0; i < 3; i++)
+        {
+            quat.at<double>(i + 1) = a * rotvec.at<double>(i);
+        }
+        quat = quat / SIGN(quat.at<double>(0));
+
+        return quat;
+    }
+
+    cv::Mat quat2rotvec(cv::Mat quat)
+    {
+        cv::Mat rotvec(3, 1, CV_64FC1, cv::Scalar(0));
+        auto angle = acos(quat.at<double>(0)) * 2;
+        auto a = sqrt((1 - pow(quat.at<double>(0), 2)));
+        for(int i = 0; i < 3; i++)
+        {
+            rotvec.at<double>(i) = angle * quat.at<double>(i + 1) / a;
+        }
+
+        return rotvec;
+    }
+
+    cv::Mat angleaxis2rotvec(double angle, cv::Mat axis)
+    {
+        return axis * angle;
+    }
+
+    std::tuple<double, cv::Mat> rotvec2angleaxis(cv::Mat rotvec)
+    {
+        cv::Mat axis(3, 1, CV_64FC1, cv::Scalar(0));
+        double angle = cv::norm(rotvec);
+        axis = rotvec / cv::norm(rotvec);
+
+        return std::make_tuple(angle, axis);
+    }
+
+    std::tuple<double, cv::Mat> quat2angleaxis(cv::Mat quat)
```

Comment:
This conversion seems 'alone'. For all others, you have made the reverse conversion:
RPY >< R ; R >< Q ; Q >< RV ; RV >< AA. I would either remove this function since one can do Q > RV > AA or add AA > Q - but then one might ask why not make all other conversions, which I believe would be to much code. We should provide a minimum required for someone to get their conversions.

Reply:
I'd add this:
    cv::Mat angleaxis2quat(double angle, cv::Mat axis)
    {
        cv::Mat quat(4, 1, CV_64FC1, cv::Scalar(0));

        quat = rotvec2quat(angleaxis2rotvec(angle, axis));
        
        return axis * angle;
    }

Reply:
I would also agree that the function may not be required. However, I thought it'd be nice to show. What I'm thinking is that it should either exist as a function, or in the _main(). If goal is to reduce size of _main(), then it would fall out through refactoring.